File: 02/02.pl
:- use_module(library(lists)).
:- use_module(library(readutil)).

move(rock).
move(paper).
move(scissors).

result(loss).
result(draw).
result(win).

play(A, A, draw).
play(rock, paper, loss).
play(rock, scissors, win).
play(paper, scissors, loss).
play(paper, rock, win).
play(scissors, rock, loss).
play(scissors, paper, win).

value(rock, 1).
value(paper, 2).
value(scissors, 3).
value(loss, 0).
value(draw, 3).
value(win, 6).

score(OpponentMove, Move, Result, Score) :-
    play(Move, OpponentMove, Result),
    value(Move, MoveScore),
    value(Result, RoundScore),
    Score is MoveScore + RoundScore.

part1_solve(List, TotalScore) :-
    maplist(part1_score, List, Scores),
    sumlist(Scores, TotalScore).

part1_score([OpponentMove, Move, _], Score) :-
    score(OpponentMove, Move, _, Score).

part2_solve(List, TotalScore) :-
    maplist(part2_score, List, Scores),
    sumlist(Scores, TotalScore).

part2_score([OpponentMove, _, Result], Score) :-
    score(OpponentMove, _, Result, Score).

parse(_, "A", rock).
parse(_, "B", paper).
parse(_, "C", scissors).
parse(part1, "X", rock).
parse(part1, "Y", paper).
parse(part1, "Z", scissors).
parse(part2, "X", loss).
parse(part2, "Y", draw).
parse(part2, "Z", win).

read_input(List) :-
    open("input_full.txt", read, Input),
    read_file_to_list(Input, Lines),
    maplist(parse_line, Lines, List).

parse_line(Line, [OpponentMove, Move, Result]) :-
    split_string(Line, " ", "", [L, R]),
    parse(_, L, OpponentMove),
    parse(part1, R, Move),
    parse(part2, R, Result).

main :-
    read_input(List),
    part1_solve(List, Part1Total),
    part2_solve(List, Part2Total),
    print(Part1Total), nl,
    print(Part2Total), nl,
    halt.

read_file_to_list(Stream, []) :-
    at_end_of_stream(Stream).

read_file_to_list(Stream, [X|L]) :-
    \+ at_end_of_stream(Stream),
    read_line_to_string(Stream, X),
    read_file_to_list(Stream, L).

File: 02/02.hs
data Move = Rock | Paper | Scissors deriving (Enum, Eq, Show)
data Result = Loss | Draw | Win deriving (Enum, Show)
data Variable = X | Y | Z deriving (Enum, Read, Show)

beats :: Move -> Move
beats move = [Scissors, Rock, Paper] !! fromEnum move

loses :: Move -> Move
loses move = [Paper, Scissors, Rock] !! fromEnum move

score :: Move -> Move -> Int
score a b | beats a == b = 6 + fromEnum a + 1
          | a == b       = 3 + fromEnum a + 1
          | otherwise    = 0 + fromEnum a + 1

fix :: Move -> Result -> Move
fix move result = [beats move, move, loses move] !! fromEnum result

parseMove :: String -> Move
parseMove "A" = Rock
parseMove "B" = Paper
parseMove "C" = Scissors

parseInput :: String -> [(Move, Variable)]
parseInput = map ((\[move, var] -> (parseMove move, read var)) . words) . lines

convert :: (Enum a, Enum b) => a -> b
convert = toEnum . fromEnum

main :: IO ()
main = do
  input <- parseInput <$> readFile "input_full.txt"
  print . sum . map (\(a, b) -> score (convert b) a) $ input
  print . sum . map (\(a, b) -> score (fix a (convert b)) a) $ input
